pgjdbc/src/main/java/org/postgresql/core/v3/BatchedQuery.java

```diff
@@ -0,0 +1,152 @@
+      blocks[index] = bq;
+    }
+    return bq;
+  }
+
+  @Override
+  public int getBatchSize() {
+    return batchSize;
+  }
+
+  /**
+   * Method to return the sql based on number of batches. Skipping the initial
+   * batch.
+   */
+  @Override
+  String getNativeSql() {
+    if (sql != null) {
+      return sql;
+    }
+    // dynamically build sql with parameters for batches
+    String nativeSql = super.getNativeSql();
+    int batchSize = getBatchSize();
+    if (batchSize < 2) {
+      sql = nativeSql;
+      return sql;
+    }
+    if (nativeSql == null) {
+      sql = "";
+      return sql;
+    }
+    int valuesBlockCharCount = 0;
+    // Split the values section around every dynamic parameter.
+    int[] bindPositions = getNativeQuery().bindPositions;
+    int[] chunkStart = new int[1 + bindPositions.length];
+    int[] chunkEnd = new int[1 + bindPositions.length];
+    chunkStart[0] = valuesBraceOpenPosition;
+    chunkEnd[0] = bindPositions[0];
+    // valuesBlockCharCount += chunks[0].length;
+    valuesBlockCharCount += chunkEnd[0] - chunkStart[0];
+    for (int i = 0; i < bindPositions.length; i++) {
+      int startIndex = bindPositions[i] + 2;
+      int endIndex =
+          i < bindPositions.length - 1 ? bindPositions[i + 1] : valuesBraceClosePosition + 1;
+      for (; startIndex < endIndex; startIndex++) {
+        if (!Character.isDigit(nativeSql.charAt(startIndex))) {
+          break;
+        }
+      }
+      chunkStart[i + 1] = startIndex;
+      chunkEnd[i + 1] = endIndex;
+      // valuesBlockCharCount += chunks[i + 1].length;
+      valuesBlockCharCount += chunkEnd[i + 1] - chunkStart[i + 1];
+    }
+    int length = nativeSql.length();
+    //valuesBraceOpenPosition + valuesBlockCharCount;
+    length += NativeQuery.calculateBindLength(bindPositions.length * batchSize);
+    length -= NativeQuery.calculateBindLength(bindPositions.length);
+    length += (valuesBlockCharCount + 1 /*comma*/) * (batchSize - 1 /* initial sql */);
+
+    StringBuilder s = new StringBuilder(length);
+    // Add query until end of values parameter block.
+    s.append(nativeSql, 0, valuesBraceClosePosition + 1);
+    int pos = bindPositions.length + 1;
+    for (int i = 2; i <= batchSize; i++) {
+      s.append(',');
+      s.append(nativeSql, chunkStart[0], chunkEnd[0]);
+      for (int j = 1; j < chunkStart.length; j++) {
+        NativeQuery.appendBindName(s, pos++);
+        s.append(nativeSql, chunkStart[j], chunkEnd[j]);
+      }
+    }
+    // Add trailing content: final query is like original with multi values.
+    // This could contain "--" comments, so it is important to add them at end.
+    s.append(nativeSql, valuesBraceClosePosition + 1, nativeSql.length());
+    sql = s.toString();
+    assert s.length() == length
+        : "Predicted length != actual: " + length + " !=" + s.length();
+    return sql;
+  }
+
+  @Override
+  public String toString() {
+    return getNativeSql();
```

Comment:
> To come back to the current situation, should it show the latest bindings only?

Ah, we can end up with the same query object and different bind parameters in the same series of batch inserts.

Let's decide on that later then.

The idea with `toSting` is external profilers that instrument pgjdbc code to measure time and report sql/binds should have some sensible way of accessing query and binds without too much hard-coding for a specific pgjdbc version.
I do support an in-house profiler, and it is not a nice thing to reimplement "support of pgjdbc" as it gets upgraded. That is why I tried to rely on simple `toString`.

It might be time to introduce some proper API (that will cover sql, binds, batches, etc)